service/config_provider.go

```diff
@@ -68,50 +68,112 @@ type ConfigProvider interface {
+// returned by NewConfigProvider()
+type ConfigProviderOption func(opts *configProvider)
+
+// WithConfigMapProvider appends to the default available providers.
+// This provider overwrites any existing provider with the same scheme.
+func WithConfigMapProvider(mp config.MapProvider) ConfigProviderOption {
+	return func(opts *configProvider) {
+		opts.configMapProviders[mp.Scheme()] = mp
+	}
+}
+
+// WithConfigMapConverters overwrites the default converters.
+func WithConfigMapConverters(c []config.MapConverterFunc) ConfigProviderOption {
+	return func(opts *configProvider) {
+		opts.configMapConverters = c
+	}
+}
+
+// withConfigUnmarshaler overwrites the default unmarshaler.
+// This func is private because providing custom ConfigUnmarshaler is not necessary since users can wrap/implement
+// ConfigProvider if needed to change the resulted config.
+func withConfigUnmarshaler(c configunmarshaler.ConfigUnmarshaler) ConfigProviderOption {
+	return func(opts *configProvider) {
+		opts.configUnmarshaler = c
+	}
+}
+
+// Deprecated: [v0.48.0] use NewConfigProvider instead.
 func MustNewConfigProvider(
 	locations []string,
 	configMapProviders map[string]config.MapProvider,
 	cfgMapConverters []config.MapConverterFunc,
 	configUnmarshaler configunmarshaler.ConfigUnmarshaler) ConfigProvider {
+	opts := []ConfigProviderOption{
+		WithConfigMapConverters(cfgMapConverters),
+		withConfigUnmarshaler(configUnmarshaler),
+	}
+	for _, c := range configMapProviders {
+		opts = append(opts, WithConfigMapProvider(c))
+	}
+	cfgProvider, err := NewConfigProvider(locations, opts...)
+	if err != nil {
+		panic(err)
+	}
+	return cfgProvider
+}
+
+// NewConfigProvider returns a new ConfigProvider that provides the configuration:
+// * Retrieve the config.Map by merging all retrieved maps from the given `locations` in order.
+// * Then applies all the config.MapConverterFunc in the given order.
+// * Then unmarshals the config.Map into the service Config.
+func NewConfigProvider(locations []string, opts ...ConfigProviderOption) (ConfigProvider, error) {
+	if len(locations) == 0 {
+		return nil, fmt.Errorf("cannot create ConfigProvider: no locations provided")
+	}
 	// Safe copy, ensures the slice cannot be changed from the caller.
 	locationsCopy := make([]string, len(locations))
 	copy(locationsCopy, locations)
-	return &configProvider{
+
+	opts = append([]ConfigProviderOption{
+		WithConfigMapProvider(filemapprovider.New()),
+		WithConfigMapProvider(envmapprovider.New()),
+		WithConfigMapProvider(yamlmapprovider.New()),
+	}, opts...)
+
+	provider := configProvider{
 		locations:          locationsCopy,
-		configMapProviders: configMapProviders,
-		cfgMapConverters:   cfgMapConverters,
-		configUnmarshaler:  configUnmarshaler,
-		watcher:            make(chan error, 1),
+		configMapProviders: make(map[string]config.MapProvider),
+		configMapConverters: []config.MapConverterFunc{
+			configmapprovider.NewExpandConverter(),
```

Comment:
This drops the `NewOverwritePropertiesConverter`.  While equivalent functionality may be available from the `yamlmapprovider` package I would expect this to surprise anyone using the `--set` flag with a suddenly non-functioning deployment.

Reply:
The `NewOverwritePropertiesConverter` cannot be added here, because it needs the values for `--set`. This is not a regression and not a "non-functioning" deployment, because:
* If users use our distros we take care of that in NewWindowsService and NewCommand;
* If users use a custom distro that does not use NewWindowsService and/or NewCommand and they use:
  *  `MustNewConfigProvider` - that did not have support for `--set` callers needed to specifically use `NewOverwritePropertiesConverter`; When they transition nothing to do;
  *  `MustNewDefaultConfigProvider` that did have support for `--set`; When transition callers need to specifically use `NewOverwritePropertiesConverter` for that;

Reply:
We need to commit to maintaining support for `--set`, then, because removal of that flag will break existing deployments.  We cannot decide that having two ways to do the same thing is wrong and remove it.

Reply:
Who we?

Reply:
The flag is not removed as pointed. Not sure what you are talking, sorry.

Reply:
How is this discussion relevant to this PR?